lib/protos.js

```diff
@@ -109,30 +125,60 @@ exports.track = function(track){};
  * // => []
  */
 
-exports.map = function(events, event){
-  var normalizedEvent = normalize(event);
-
-  return foldl(function(matchingEvents, val, key, events) {
-    // If true, this is a `mixed` value, which is structured like so:
-    //     { key: 'testEvent', value: { event: 'testEvent', someValue: 'xyz' } }
-    // We need to extract the key, which we use to match against
-    // `normalizedEvent`, and return `value` as part of `matchingEvents` if that
-    // match succeds.
-    if (type(events) === 'array') {
-      // If there's no key attached to this event mapping (unusual), skip this
-      // item.
-      if (!val.key) return matchingEvents;
-      // Extract the key and value from the `mixed` object.
-      key = val.key;
-      val = val.value;
+exports.map = function(mapping, key){
+  var normalizedComparator = normalize(key);
+  var mappingType = getMappingType(mapping);
+
+  function getMappingType(mapping) {
+    if (type(mapping) === 'object' && every(isString, mapping)) return 'map';
+    if (type(mapping) === 'array' && every(isString, mapping)) return 'array';
+    if (type(mapping) === 'array' && every(isMixed, mapping)) return 'mixed';
+    return 'unknown';
+  }
+
+  function isString(item) {
+    return type(item) === 'string';
+  }
+
+  /* eslint-disable no-extra-parens */
+  function isMixed(item) {
+    return (
+      type(item) === 'object'
+      && has.call(item, 'key')
+      && has.call(item, 'value')
```

Comment:
This will still be affected by the bug detailed in [this comment](https://github.com/segmentio/analytics.js-integration/pull/45#discussion_r32876462).

Not really sure under which conditions it occurs, but when Amir implemented initially it he included that check, so I'm assuming it's important. (We really should add a test for this, too.)